std/string.d

```diff
@@ -2937,26 +2937,96 @@ if (isForwardRange!Range && isSomeChar!(ElementEncodingType!Range) &&
 {
     static import std.ascii;
     static import std.uni;
-    import std.utf : decodeFront;
 
-    while (!input.empty)
+    static if (is(Unqual!(ElementEncodingType!Range) == dchar)
+        || is(Unqual!(ElementEncodingType!Range) == wchar))
     {
-        auto c = input.front;
-        if (std.ascii.isASCII(c))
+        // Decoding is never needed for dchar. It happens not to be needed
+        // here for wchar because no whitepace is outside the basic
+        // multilingual plane meaning every whitespace character is encoded
+        // with a single wchar and due to the design of UTF-16 those wchars
+        // will not occur as part of the encoding of multi-wchar codepoints.
+        static if (isDynamicArray!Range || (isRandomAccessRange!Range && hasLength!Range
+            && __traits(compiles, { size_t s = input.length / 2; input = input[s .. $]; })))
         {
-            if (!std.ascii.isWhite(c))
-                break;
-            input.popFront();
+            size_t i = 0;
+            for (const size_t end = input.length; i < end; ++i)
+            {
+                if (!std.uni.isWhite(input[i]))
+                    break;
+            }
+            input = input[i .. $];
+            return input;
```

Comment:
How about:
```D
foreach(i; 0 .. input.length)
{
    if (!std.uni.isWhite(input[i]))
        return input[i .. $];
}
return input[$ .. $];
```

Reply:
Seem to be nearly the same, so I'll use your suggestion.